### PR TITLE
[DOC] introducing sphinxcontrib-Bibtex to improve reference management

### PR DESCRIPTION
**Pull request by skoudoro**
> build succeeded, 1871 warnings.

A lot of those warning are due to references. This PR add `sphinxcontrib-bibtex` to manage references, avoid duplication, avoid formatting error, etc. 

This PR will take a bit of time to be finalize so I wonder if I should split it in multiple PR or not.  It should make review process easier. Looking forward for your feedback.

### CODE REVIEW

**Comment by arokem** [comment]
Splitting it up makes sense to me. There's no reason to wait until it's all done to merge these additions, right?

**Comment by skoudoro** [comment]
> There's no reason to wait until it's all done to merge these additions, right?

right, so you can merge, and I will do that step by step starting with the examples and then the docstring in the codebase

File: dipy/reconst/dki.py
Comment (arokem) [inline dipy/reconst/dki.py:386]: This is a bit of an eyesore. Maybe we could we use a uniform format for bibtex keys? For example, "NetoHenriques2015NeuroImage"

**Review by arokem — COMMENTED** [review]
Looks great! What do you think about normalizing the bibtex keys?

File: dipy/reconst/dki.py
Comment (skoudoro) [reply dipy/reconst/dki.py:386]: > This is a bit of an eyesore

ahahah, you are right

> Maybe we could we use a uniform format for bibtex keys?

I agree, even if the bibtex file is big. I like the beginning of your example but I do not think we should finish with the journal. so just {firstname}{name}{year}{letter | optional if multiple paper} is that ok ?

Comment (arokem) [reply dipy/reconst/dki.py:386]: Yeah. That would work well.

File: doc/diffusion.bib
Comment (jhlegarreta) [inline doc/diffusion.bib:322]: This symbol and other similar ones introduced are not plain, BIbTeX symbols; e.g. the correct BibTeX spelling for Evren's surname would be `\"{O}zarslan`. I can have a go to the file to fix them. I assume the changes were done inadvertently introduced when running some tool.

File: doc/diffusion.bib
Comment (jhlegarreta) [inline doc/diffusion.bib:329]: Similar to the above, but in this case the hyphen does exist in BibTeX; usually BibTeX is smart enough to change it to the en dash for page ranges.

**Review by jhlegarreta — CHANGES_REQUESTED** [review]
Thanks for doing this @skoudoro.

Not sure if `sphinxcontrib-bibtex` is actively maintained, but I do agree that this or other similar mechanism is required.

Some comments in-line.

**Comment by jhlegarreta** [comment]
I agree to the conversation in https://github.com/dipy/dipy/pull/2810#discussion_r1191500875. I am sure there is also a way to generate the keys automatically using a self-defined pattern, but we may leave that as a separate task/PR.

Also, making all entries consistent (fields, names, etc.) would ease reading the file (e.g. when maintainers need to go through it). Something that might be partially automated? Probably best as a separate PR.

File: doc/diffusion.bib
Comment (jhlegarreta) [reply doc/diffusion.bib:322]: I can have a look at this and amend the commit.

**Comment by jhlegarreta** [comment]
Also, If BibTeX references will be/are split across modules (e.g. segment, tracking, docs, etc.), it may be harder to maintain, as the same reference may fall into different categories, duplicated entries may exist (unless the automated tool prevents them, but then it may not be evident where a reference should belong to).

**Comment by jhlegarreta** [comment]
I had forgotten that we had started this. Thanks for reviving it :+1:. I had done a partial local test during the last few days, and the tool seems to be working as expected. 

Re: https://github.com/dipy/dipy/pull/2810#pullrequestreview-1432731863
> Not sure if sphinxcontrib-bibtex is actively maintained, but I do agree that this or other similar mechanism is required.

There has been a release recently, so it looks like it is.

Other thoughts:
- The current bibtex is long and hard to read (layout, contains the abstracts, and lots of other additional fields, etc.). It was generated by some bibliography tool, and unless we keep the underlying file/collection (whatever that is) before the exportation in a shared repository, it will be hard to keep up with that. And even in that case, whenever a ref needs to be changed or added, we would depend on a separate repository. So manually adding refs does not look that hard (the time-consuming part is the current transition).
- It is also hard to locate references in the current file as there does not seem to be any sorting; having a type (article, inproceedings) and alphabetical (key -author surname, year) sorting would help a lot IMO.
- So I would clearly remove the current references.
- I have gathered a significant amount of references (close to 150) as I've been going through the doc warnings and best would be to add them as an additional commit to the bibtex file. Having them would save time and get us closer to fixing the doc issues. In the same commit, I can then remove the existing refs in the bibtex file (many of the refs do not appear at all across the code base, some are incomplete, and as noted, they have other problems).

Let me know what you think about adding the refs I've gathered @skoudoro.

Re https://github.com/dipy/dipy/pull/2810#issuecomment-1543985963

> I will do that step by step starting with the examples and then the docstring in the codebase

- The quickest ones to fix (as there are very few files and refs) would be the ones in the `theory` doc module, and maybe at that point the PR can be already merged, to then address the examples, as you say, then other modules as time permits. I can event push that as I tested on these locally.

A few other things to note/test:
  - Use of `.. bibliography::` vs `.. footbibliography::` (and `:cite:t:`, `:cite:p:` vs the `footcite` counterparts).
  - I've read that there needs to be a single directive across the entire code base (not sure/not tested).
  - We also have footnotes (different from notes) that `sphinxcontrib-bibtex` can manage.
  - Cross-refs in figure captions may need to be tested.


I have another batch of doc warning fixes that I will push in the coming days so that we can further decrease the number and be closer to having the citation-related ones only (but I also see in the API that blocks that do not contain warnings have at times formatting problems, especially when math blocks are involved, but that will be for another round of PRs).

**Comment by skoudoro** [comment]
> I had forgotten that we had started this. Thanks for reviving it 👍. I had done a partial local test during the last few days, and the tool seems to be working as expected.

thanks 👍 

> There has been a release recently, so it looks like it is.

yep, it is and it seems it does not need a lot of update

> The current bibtex is long and hard to read.......It is also hard to locate references in the current file.......So I would clearly remove the current references.

I agree with this statement. For now, do you want me to just rename it and then, when we have the new file, we can remove it?

> I have gathered a significant amount of references (close to 150) 

Please, create a PR, after merging your new references, I can rebase this one and start the proper way.

> Use of .. bibliography::

Use of `.. bibliography::` needs to be use only one time. That why we should use `.. footbibliography::` in priority. I need to read again the documentation, it was a long time again. the `mne-python` project use it the same way if you have time to have a look

> I've read that there needs to be a single directive across the entire code base (not sure/not tested).

yes, and that is the `.. bibliography::`.   see my comment just above.

> We also have footnotes (different from notes) that sphinxcontrib-bibtex can manage.

another steps after fixing bibliography

> I have another batch of doc warning fixes that I will push in the coming days so that we can further decrease the number and be closer to having the citation-related ones only

ok 👍 



Thank you for looking into this @jhlegarreta

**Comment by jhlegarreta** [comment]
@skoudoro OK, thanks for the feedback. See PR #3308 for the bibliography file:
- There may be a few references missing, but tried to go over all files to gather all references.
- @arokem @skoudoro Simplified the key convention to `lastNameYear(Letter)`. IMO it should be enough.

**Comment by jhlegarreta** [comment]
Did not dare to touch the `doc/diffusion.bib` Serge, but I guess we would be good rebasing on `master`, and fixing the conflicts by dropping the changes in here.

I will slowly work on the transition of the references in the core modules in a separate PR.

**Comment by skoudoro** [comment]
Hi @jhlegarreta,

I did the rebase. However, `sphinxcontrib-bibtex` is failing due to many unknown entries.  (like `@electronics`, `@dataset`)

Indeed, they do not exist when I look at this page: https://www.bibtex.com/e/entry-types/.

Can you tell me where they come from and which tool is recognizing these types of entries? Do you want to create a PR to simplify that or should I go ahead here?

Thank you for your feedback

**Comment by jhlegarreta** [comment]
Sorry for the headaches this might have caused @skoudoro: PR #3312.

File: doc/references.bib
Comment (jhlegarreta) [inline doc/references.bib:828]: {{Optimal strategies for measuring diffusion in anisotropic systems by magnetic resonance imaging}},
```suggestion
  title     = {{Optimal strategies for measuring diffusion in anisotropic systems by magnetic resonance imaging}},
```

File: doc/references.bib
Comment (jhlegarreta) [inline doc/references.bib:829]: Magnetic Resonance in Medicine
```suggestion
  journal   = {Magnetic Resonance in Medicine},
```

File: doc/references.bib
Comment (jhlegarreta) [inline doc/references.bib:827]:
```suggestion
  author    = {Derek K. Jones and Mark A. Horsfield and Andrew Simmons},
```

**Review by jhlegarreta — COMMENTED** [review]
Sorry, I missed this one.

**Comment by skoudoro** [comment]
> Sorry for the headaches this might have caused @skoudoro: PR https://github.com/dipy/dipy/pull/3312.

I have to admit, not easy to find the issue 😅 

all back to normal, I think this PR is ready to go

**Review by jhlegarreta — COMMENTED** [review]
:+1: thanks for doing this: the number of warnings has decreased in the documentation CI builds.

I would have preferred to see this
https://github.com/dipy/dipy/pull/2810/files#diff-c1e15eead4a02baaf8049c7bc83badef5892399b556e74a6e35a84c5926b0a33L343-L344

as a separate commit. If you are not willing to change that, and if your local tests have shown that we are safe to make the transition across the code base, let me know and I will merge as is.

Edit: thanks for having fixed a number of oversights of mine in `references.bib`.

**Comment by skoudoro** [comment]
> I would have preferred to see this

Sorry for that... if you have a quick instruction line to change that, I can do it, otherwise, I am not willing to spend time on this. Sorry. 

>  if your local tests have shown that we are safe to make the transition across the code base, let me know

Yes, it is safe. you can go ahead ! see some screenshot below

image
image
image

Now, we need to update all our codebase (docstring, rst and examples). Let's start with docstring and/or to avoid conflicting with #3306.

**Comment by jhlegarreta** [comment]
> Sorry for that... if you have a quick instruction line to change that, I can do it, otherwise, I am not willing to spend time on this. Sorry.

:+1: Sounds good.

> Yes, it is safe. you can go ahead ! see some screenshot below

Looks :100:. Merging.

**Review by jhlegarreta — APPROVED** [review]

File: doc/diffusion.bib
Comment (jhlegarreta) [reply doc/diffusion.bib]: Removed existing entries were removed in PR #3308. Resolving this.

File: doc/diffusion.bib
Comment (jhlegarreta) [reply doc/diffusion.bib]: Removed existing entries were removed in PR #3308. Resolving this.

**Pull request merged**
